File: python/customer/adjust_internal_certificate.py
from ultracart.apis import CustomerApi
from samples import api_client
from ultracart.models import AdjustInternalCertificateRequest

# Create the customer API instance
customer_api = CustomerApi(api_client())

# Set the email and retrieve customer
email = "test@ultracart.com"
customer = customer_api.get_customer_by_email(email).customer
customer_oid = customer.customer_profile_oid

# Create adjustment request
adjust_request = AdjustInternalCertificateRequest(
    description="Adjusting customer cashback balance because they called and complained about product.",
    expiration_days=365,  # expires in 365 days
    vesting_days=45,  # customer has to wait 45 days to use it
    adjustment_amount=59,  # add 59 to their balance
    order_id='DEMO-12345',  # or leave None. This ties the adjustment to a particular order
    entry_dts=None  # use current time
)

# Adjust internal certificate
api_response = customer_api.adjust_internal_certificate(customer_oid, adjust_request)

# Check for errors
if api_response.error is not None:
    print(f"Developer Message: {api_response.error.developer_message}")
    print(f"User Message: {api_response.error.user_message}")
    exit()

# Print response details
print(f"Success: {api_response.success}")
print(f"Adjustment Amount: {api_response.adjustment_amount}")
print(f"Balance Amount: {api_response.balance_amount}")

# Optional: full response inspection
print(api_response)

File: typescript/customer/AdjustInternalCertificate.ts
import {customerApi} from '../api';
import {
    Customer,
    AdjustInternalCertificateRequest,
    AdjustInternalCertificateResponse
} from 'ultracart_rest_api_v2_typescript';
import {DateTime} from 'luxon';

export class AdjustInternalCertificate {
    /**
     * Adjusts the cashback balance of a customer. This method's name is adjustInternalCertificate, which
     * is a poor choice of naming, but results from an underlying implementation of using an internal gift certificate
     * to track cashback balance. Sorry for the confusion.
     *
     * This method requires a customer profile oid. This is a unique number used by UltraCart to identify a customer.
     * If you do not know a customer's oid, call getCustomerByEmail() to retrieve the customer and their oid.
     *
     * Possible Errors:
     * Missing adjustment amount -> "adjust_internal_certificate_request.adjustment_amount is required and was missing"
     */
    public static async execute(): Promise<void> {
        const email: string = "test@ultracart.com";

        // Retrieve customer by email
        const customerResponse = await customerApi.getCustomerByEmail({email});
        const customer: Customer | undefined = customerResponse.customer;

        if (!customer || !customer.customer_profile_oid) {
            throw new Error("Customer not found or missing customer profile OID");
        }

        const customerOid: number = customer.customer_profile_oid;

        const adjustRequest: AdjustInternalCertificateRequest = {
            description: "Adjusting customer cashback balance because they called and complained about product.",
            expiration_days: 365, // expires in 365 days
            vesting_days: 45, // customer has to wait 45 days to use it.
            adjustment_amount: 59, // add 59 to their balance.
            order_id: "DEMO-12345", // or leave undefined. this ties the adjustment to a particular order.
            entry_dts: DateTime.now().setZone('America/New_York').toISO() // use current time in ISO format
        };

        const apiResponse: AdjustInternalCertificateResponse = await customerApi.adjustInternalCertificate({
            customerProfileOid: customerOid,
            adjustInternalCertificateRequest: adjustRequest
        });

        if (apiResponse.error) {
            console.error(apiResponse.error.developer_message);
            console.error(apiResponse.error.user_message);
            throw new Error("Failed to adjust internal certificate");
        }

        console.log(`Success: ${apiResponse.success}`);
        console.log(`Adjustment Amount: ${apiResponse.adjustment_amount}`);
        console.log(`Balance Amount: ${apiResponse.balance_amount}`);

        console.log(apiResponse);
    }
}